### ISSUE TITLE
cross compile issue on ubuntu16 

### ISSUE THREAD

**Issue by bb33bb** [openssl/openssl#12380]
git clone https://github.com/openssl/openssl
cd openssl
mkdir build
cd build
../Configure mingw64 --prefix=/usr/lib/win/openssl no-asm no-shared

error like this

bj -c -o providers/fips/fips-dso-self_test_kats.obj ../providers/fips/self_test_kats.c
/usr/bin/perl ../util/mkdef.pl --ordinals ../util/providers.num  --name providers/fips --OS mingw > providers/fips.def
Use of uninitialized value $libname in concatenation (.) or string at ../util/mkdef.pl line 284.
Use of uninitialized value $libname in concatenation (.) or string at ../util/mkdef.pl line 284.
x86_64-w64-mingw32-gcc -m64 -Wall -O3 -static-libgcc -shared -Wl,--enable-auto-image-base   \
	-o providers/fips.dll providers/fips.def \
	providers/fips/fips-dso-fipsprov.obj \
	providers/fips/fips-dso-self_test.obj \
	providers/fips/fips-dso-self_test_kats.obj \
	providers/libimplementations.a providers/libcommon.a providers/libfips.a -lws2_32 -lgdi32 -lcrypt32 
/usr/bin/x86_64-w64-mingw32-ld: providers/fips.def:5: syntax error
/usr/bin/x86_64-w64-mingw32-ld:providers/fips.def: file format not recognized; treating as linker script
/usr/bin/x86_64-w64-mingw32-ld:providers/fips.def:2: syntax error
collect2: error: ld returned 1 exit status
Makefile:14304: recipe for target 'providers/fips.dll' failed
make[1]: *** [providers/fips.dll] Error 1
make[1]: Leaving directory '/home/avboy/Desktop/openssl/build'
Makefile:2666: recipe for target 'build_sw' failed
make: *** [build_sw] Error 2

**Comment by kaduk** [comment]
> ../Configure mingw64 --prefix=/usr/lib/win/openssl no-asm no-shared

Are you copying this line from some directions somewhere?  It is highly unusual if you are running on an ubuntu system, as mingw64 is a target that runs on a Windows environment.

**Comment by frustreated** [comment]
> > ../Configure mingw64 --prefix=/usr/lib/win/openssl no-asm no-shared
> 
> Are you copying this line from some directions somewhere? It is highly unusual if you are running on an ubuntu system, as mingw64 is a target that runs on a Windows environment.

openssl version 1.xxx compiles ok
personal need to do like this . another project based on openssl

**Comment by mattcaswell** [comment]
When cross compiling you are supposed to configure using the option `--cross-compile-prefix` as documented in the INSTALL file (this was the case with 1.1.x too):

https://github.com/openssl/openssl/blob/master/INSTALL.md#cross-compile-prefix

I don't have any ubuntu16 version, but the following works for me on ubuntu 20.04:

````
perl Configure --cross-compile-prefix=/usr/bin/i686-w64-mingw32- mingw64
````

**Comment by levitte** [comment]
I would like to see the output of `perl ./configdata.pm --dump`.  Since OpenSSL is built as a sub-project for something else, it might be that some environment variable is playing unexpected tricks

**Comment by frustreated** [comment]
> When cross compiling you are supposed to configure using the option `--cross-compile-prefix` as documented in the INSTALL file (this was the case with 1.1.x too):
> 
> https://github.com/openssl/openssl/blob/master/INSTALL.md#cross-compile-prefix
> 
> I don't have any ubuntu16 version, but the following works for me on ubuntu 20.04:
> 
> ```
> perl Configure --cross-compile-prefix=/usr/bin/i686-w64-mingw32- mingw64
> ```

using command 

perl Configure --cross-compile-prefix=/usr/bin/i686-w64-mingw32- mingw64

非常伤心。
cc1: sorry, unimplemented: 64-bit mode not compiled in

**Comment by mattcaswell** [comment]
Sorry, my mistake the actual Configure line is:

````
perl Configure -d --cross-compile-prefix=/usr/bin/x86_64-w64-mingw32- mingw64
````

**Comment by frustreated** [comment]
> Sorry, my mistake the actual Configure line is:
> 
> ```
> perl Configure -d --cross-compile-prefix=/usr/bin/x86_64-w64-mingw32- mingw64
> ```

老哥牛！确实搞定了。继续编译。

**Comment by frustreated** [comment]
没问题的话明天关闭issue

**Comment by nhorman** [comment]
Marking as inactive to be closed at the end of 3.4 dev, barring further input
